### PR TITLE
RSDK-1960, RSDK-539 Test entire config flow

### PR DESCRIPTION
**Pull request by maximpertsov**
Stub out network calls and test the entire config reader flow

### CODE REVIEW

File: config/reader.go [hunk 1]
```diff
@@ -126,37 +126,33 @@ func clearCache(id string) {
 	})
 }
 
-func readCertificateDataFromCloudGRPC(ctx context.Context,
+type remoteReader struct {
+	client apppb.RobotServiceClient
+}
```
Comment (maximpertsov) [inline config/reader.go:131]: pulling out this type lets us inject a mock robot service client

File: config/reader.go [hunk 2]
```diff
@@ -384,8 +393,13 @@ func fromReader(
...
-	if shouldReadFromCloud && cfgFromDisk.Cloud != nil {
-		cfg, err := readFromCloud(ctx, cfgFromDisk, nil, true, true, logger)
+	if rrFactory != nil && cfgFromDisk.Cloud != nil {
+		rr, closeFunc, err := rrFactory(ctx, cfgFromDisk.Cloud, logger)
+		if err != nil {
+			return nil, err
+		}
+		defer utils.UncheckedErrorFunc(closeFunc)
+		cfg, err := rr.readFromCloud(ctx, cfgFromDisk, nil, true, true, logger)
 		return cfg, err
```
Comment (maximpertsov) [inline config/reader.go:403]: this is the injection point. 

having a factory argument is ugly... basically we need to read some information from the config on disk to create a grpc connection to app (robot part id, secret, app address), and that reading happens here. 

curious if folks have ideas to improve this.

File: config/reader.go [hunk 3]
```diff
@@ -359,17 +368,17 @@ func FromReader(
...
+// fromReader reads a config from the given reader and specifies
 // where, if applicable, the file the reader originated from.
 func fromReader(
 	ctx context.Context,
 	originalPath string,
 	r io.Reader,
 	logger logging.Logger,
-	shouldReadFromCloud bool,
+	rrFactory remoteReaderFactory,
 ) (*Config, error) {
```
Comment (maximpertsov) [inline config/reader.go:382]: this function is the injection point. 

having a factory argument is ugly... basically we need to read some information from the config on disk to create a grpc connection to app (robot part id, secret, app address), and that reading happens here. 

curious if folks have ideas to improve this.

File: testutils/injectpb/robot_service_client.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package injectpb
```
Comment (maximpertsov) [inline testutils/injectpb/robot_service_client.go:1]: we can't add this to the `inject` package due to circular dependencies

File: config/watcher.go [hunk 5]
```diff
@@ -56,6 +57,16 @@ func newCloudWatcher(ctx context.Context, config *Config, logger logging.Logger)
...
+	// TODO: the watcher should also restablish this connection.
+	conn, err := CreateNewGRPCClient(ctx, config.Cloud, logger)
+	if err != nil {
+		cancel()
+		logger.Warn("failed to connect to app - cannot watch for config updates")
+		return nil
+	}
+	defer utils.UncheckedErrorFunc(conn.Close)
+	rr := remoteReader{apppb.NewRobotServiceClient(conn)}
+
```
Comment (maximpertsov) [inline config/watcher.go:69]: this is a significant behavior change - before we were creating connections on demand closing them within the loop - here we just create a connection once. the biggest question is how we handle a connection failure.

File: config/reader.go (hunk 3 above)
Comment (cheukt) [reply config/reader.go:382]: i've been thinking about it a bit and not sure if there's anything easy we can do here and still allow us to test fromReader() end to end. We could possibly split fromReader up, so we can test each part separately and not have to instantiate the remote reader within the function.

but if we can't really make it less ugly, this is fine as well

File: config/watcher.go (hunk 5 above)
Comment (cheukt) [reply config/watcher.go:69]: can we nil out the client on every error and then remake the connection the next time around?

File: config/reader.go (hunk 3 above)
Comment (maximpertsov) [reply config/reader.go:382]: cool, thanks for taking the time to think about it

File: config/watcher.go [hunk 6]
```diff
@@ -66,11 +69,25 @@ func newCloudWatcher(ctx context.Context, config *Config, logger logging.Logger)
...
+			if conn == nil {
+				var err error
+				conn, err = CreateNewGRPCClient(ctx, config.Cloud, logger)
+				if err != nil {
+					conn = nil
+					logger.Warn("failed to connect to app")
+					continue
+				}
+			}
+			rr := remoteReader{apppb.NewRobotServiceClient(conn)}
```
Comment (cheukt) [inline config/watcher.go:82]: i think we should use `newRemoteReader` here too, even if that means connection management goes into the remoteReader itself

Comment (maximpertsov) [reply config/watcher.go:82]: ah yeah, not sure why I didn't use that...

**Review by cheukt — APPROVED** [review]

**Pull request merged**